add testing web clients from code (#1109)
jellyfin/jellyfin.org: docs/general/testing/web/index.md

:::caution
Using a development version of clients may lead to data corruption or loss on the server. Please use a dedicated test server and make regular backups of [Jellyfin files](/docs/general/administration/configuration#server-paths).
:::

## Testing from source code

> Jellfin Web can be hosted as a standalone application without being associated to a Jellyfin server. In that case, a separate Jellyfin server will need to be prepared.

Reviewer: ```suggestion
Jellyfin Web can be hosted as a standalone application without being associated to a Jellyfin server. In that case, a separate Jellyfin server will need to be prepared.
```

:::caution
Using a development version of clients may lead to data corruption or loss on the server. Please use a dedicated test server and make regular backups of [Jellyfin files](/docs/general/administration/configuration#server-paths).
:::

## Testing from source code

> Jellyfin web clients can be hosted as a standalone application without being associated to a Jellyfin server. In that case, a separate Jellyfin server will need to be prepared.

Reviewer: ```suggestion
Jellyfin web clients can be hosted as a standalone application without being associated with a Jellyfin server. In that case, a separate Jellyfin server will need to be prepared.
```

> Jellyfin web clients can be hosted as a standalone application without being associated with a Jellyfin server. In that case, a separate Jellyfin server will need to be prepared.

Reviewer 2: ```suggestion
Jellyfin web clients can be hosted as a standalone application without being associated with a Jellyfin server. In that case, a separate Jellyfin server will be needed.
```
I'm not sure that second sentence it's necessary as well. The first one is clear enough and I feel like the second one could be misinterpreted, leading people to think that at least 2 servers are needed somehow?

"Test server" below refers to the device hosting the web client in the test setup.

### Prerequisites

> Below are a list of things to prepare before testing Jellyfin web clients.

Reviewer: ```suggestion
Below is a list of things to prepare before testing Jellyfin web clients.
```

- Have an instance of Jellyfin Server. A dedicated testing setup is recommended.
- Install [NodeJS, npm and nvm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm) on the test server.
- Install [Git](https://github.com/git-guides/install-git) on the test server.
- (Optional) Install [GitHub CLI](https://cli.github.com/) or [GitHub Desktop](https://github.com/apps/desktop) on the test server

> - (Optional) A text editor or web IDE to make changes during testing, eg. [VSCode](https://code.visualstudio.com/), [Notepad++](https://notepad-plus-plus.org/) ,[Jetbrains Webstorm](https://www.jetbrains.com/webstorm/).

Reviewer 2: ```suggestion
- (Optional) A text editor or web IDE to make changes during testing, eg. [VSCode](https://code.visualstudio.com/), [Notepad++](https://notepad-plus-plus.org/), [Jetbrains Webstorm](https://www.jetbrains.com/webstorm/).
```

:::note
Installing Github CLI or Github Desktop will automatically install Git on the system.
:::

### Obtaining Source Code

The source code of the web clients can simply be cloned from their respective Github repositories: [Jellyfin Web](https://github.com/jellyfin/jellyfin-web/), [Jellyfin Vue](https://github.com/jellyfin/jellyfin-vue/)
Instructions on how to do so can be found [here](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository).

### Checking out branches

> The `master` branch is where the unstable build is built from. If you wish to test over there, please checkout that branch. to checkout the `master` branch, simply do the following.

Reviewer: ```suggestion
The `master` branch is where the unstable build is built from. If you wish to test over there, please checkout that branch. To checkout the `master` branch, simply do the following.
```

- Git CLI, Github CLI: open a terminal in the web folder and run `git checkout master`.

> - Github Desktop: Select the `master` branch in the branch dropdown menu.

Reviewer: ```suggestion
- GitHub Desktop: Select the `master` branch in the branch dropdown menu.
```

> ### Checking out PRs

Reviewer 2: Let's use Pull Request from the beginning here

PRs are special branches submitted by contributors with the goal of eventually being merged into `master`. To checkout a PR, please follow the instructions below.

1. Find the PR you would like to checkout in the list of open PRs: [Jellyfin Web](https://github.com/jellyfin/jellyfin-web/pulls), [Jellyfin Vue](https://github.com/jellyfin/jellyfin-vue/pulls).
2. Checkout the PR in Git: [Git CLI and Github CLI](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally), [Github Desktop](https://docs.github.com/en/desktop/working-with-your-remote-repository-on-github-or-github-enterprise/viewing-a-pull-request-in-github-desktop)

### Install Dependencies

1. Navigate to the web directory (the folder containing the cloned contents of the repository)
2. Open a terminal window in the folder.

> 3. run `npm install` to install prerequisites. This will need to be done every time you switch to a new branch.

Reviewer: ```suggestion
3. Run `npm install` to install the prerequisites. This will need to be done every time you switch to a new branch.
```

> 3. Run `npm install` to install dependencies. This will need to be done every time you switch to a new branch, as different branches might have different versions of dependencies.

Reviewer 2: I'd remove this section and link to the documentation of Web and Vue, since this is perhaps the section more likely to be changed in the future (while the others are less likely to). Things like setting up a monorepo, changing package manager, etc... Will outdate this easily. This, in fact, is not applicable to Vue (you must `cd frontend` first).

I'm also not sure the next one is needed as well for the same reasons. Only a small port change (e.g: Web switches to Vite and the new port is used instead of the old one) and this is out of date.

All this maybe could be solved by using a sentence similar to the last one applicable for Vue, but for web, linking to the repo where info is probably going to be more up to date.

I leave this up to you/other reviewers though. I also understand my concerns are more applicable to Vue in this case (since it usually live sin the bleeding edge) than Web.

### Start Development Server

1. Open a terminal window in the folder of the cloned contents.

> 2. run `npm start` to start a development server. Allow firewall access when prompted if you are using Windows. The server will start on port 8080 for Jellyfin Web and Port 3000 for Jellyfin Vue. The server will auto update if the contents of the folder is changed.

Reviewer: ```suggestion
2. Run `npm start` to start a development server. Allow firewall access when prompted if you are using Windows. The server will start on port 8080 for Jellyfin Web and Port 3000 for Jellyfin Vue. The server will auto update if the contents of the folder are changed.
```

By default, never trust any artifacts given by anyone outside the official channels if you can't inspect the source code first.
They might compromise your system or track your activity!
:::

## Testing from source code

Jellyfin web clients can be hosted as a standalone application without being associated with a Jellyfin server. In that case, a separate Jellyfin server will be needed.

"Test server" below refers to the device hosting the web client in the test setup.

### Prerequisites

Below is a list of things to prepare before testing Jellyfin web clients.

- Have an instance of Jellyfin Server. A dedicated testing setup is recommended.

> - Install [NodeJS, npm and nvm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm) on the test server.

Reviewer 2: Vue is strictly current LTS as well, yes.

:::

### Obtaining Source Code

The source code of the web clients can simply be cloned from their respective GitHub repositories: [Jellyfin Web](https://github.com/jellyfin/jellyfin-web/), [Jellyfin Vue](https://github.com/jellyfin/jellyfin-vue/)
Instructions on how to do so can be found [here](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository).

### Checking out branches

The `master` branch is where the unstable build is built from. If you wish to test over there, please checkout that branch. To checkout the `master` branch, simply do the following.

- Git CLI, GitHub CLI: open a terminal in the web folder and run `git checkout master`.
- GitHub Desktop: Select the `master` branch in the branch dropdown menu.

### Checking out pull requests

Pull requests are special branches submitted by contributors with the goal of eventually being merged into `master`. To checkout a pull request, please follow the instructions below.

1. Find the pull request you would like to checkout in the list of open pull requests: [Jellyfin Web](https://github.com/jellyfin/jellyfin-web/pulls), [Jellyfin Vue](https://github.com/jellyfin/jellyfin-vue/pulls).

> 2. Checkout the pull request in Git: [Git CLI and GitHub CLI](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally), [GitHub Desktop](https://docs.github.com/en/desktop/working-with-your-remote-repository-on-github-or-github-enterprise/viewing-a-pull-request-in-github-desktop)

Reviewer 3: ```suggestion
2. Checkout the pull request in Git: [Git CLI and GitHub CLI](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally), [GitHub Desktop](https://docs.github.com/en/desktop/working-with-your-remote-repository-on-github-or-github-enterprise/viewing-a-pull-request-in-github-desktop).
```

> ### Installing Dependencies and Starting a Development Server

Reviewer 3: The capitalization of the headings is inconsistent. Are we only capitalizing the first letter of the first word or every word?

By default, never trust any artifacts given by anyone outside the official channels if you can't inspect the source code first.
They might compromise your system or track your activity!
:::

## Testing from source code

Jellyfin web clients can be hosted as a standalone application without being associated with a Jellyfin server. In that case, a separate Jellyfin server will be needed.

"Test server" below refers to the device hosting the web client in the test setup.

### Prerequisites

Below is a list of things to prepare before testing Jellyfin web clients.

- Have an instance of Jellyfin Server. A dedicated testing setup is recommended.
- Install the latest **LTS** version of [NodeJS and npm](https://nodejs.org/en/download/) on the test server.
- Install [Git](https://github.com/git-guides/install-git) on the test server.

> - (Optional) Install nvm: [macOS, Linux](https://github.com/nvm-sh/nvm?tab=readme-ov-file#installing-and-updating), [Windows](https://github.com/coreybutler/nvm-windows)

Reviewer 3: lgtm after this :rocket: 

```suggestion
- (Optional) Install nvm: [macOS, Linux](https://github.com/nvm-sh/nvm?tab=readme-ov-file#installing-and-updating), [Windows](https://github.com/coreybutler/nvm-windows).
```

2. run `npm start` to start a development server. Allow firewall access when prompted if you are using Windows. The server will start on port 8080 for Jellyfin Web and Port 3000 for Jellyfin Vue. The server will auto update if the contents of the folder is changed.
3. Connect client devices to the server using its IP address in order to start testing. You will need to connect a server manually. To use the demo instance, enter `https://demo.jellyfin.org/unstable` in the server address field.
4. When you have finished testing, or would like to switch to a different branch, press `Ctrl+C` in the terminal window to stop the server.

> More info on how to setup Jellyfin Vue in a development environemt can be found [here](https://github.com/jellyfin/jellyfin-vue/wiki/Contributing#development-setup)

Reviewer 2: ```suggestion
More info on how to setup Jellyfin Vue in a development environemt can be found [here](https://github.com/jellyfin/jellyfin-vue/wiki/Contributing#development-setup)
```
```suggestion
More info on how to setup Jellyfin Vue in a development environmet can be found [here](https://github.com/jellyfin/jellyfin-vue/wiki/Contributing#development-setup)
```
